### PR TITLE
[onert/backend] Refactor train/Memory.test.cc

### PR DESCRIPTION
**Pull request by zetwhite**
This PR refactors MemoryPlanner.test.cc to use template verifier.
This template will be used for LayerScopeTensrIndex tests.

ONE-DCO-1.0-Signed-off-by: seunghui youn <sseung.youn@samsung.com>

related : https://github.com/Samsung/ONE/pull/13892#pullrequestreview-2278797746

### CODE REVIEW

File: runtime/onert/backend/train/MemoryPlanner.test.cc [hunk 1]
```diff
@@ -15,19 +15,88 @@
...
 using onert::ir::OperandIndex;
 using onert::ir::OperationIndex;
 
-// TODO: Add test testcase for {Bump, FirstFit, WIC}Planner<LayerScopeTensor>
+namespace
+{
+
+template <typename T> T to_index(uint32_t first, uint32_t second)
+{
+  throw std::runtime_error{"unimplemented to_index"};
```
Comment (ragmani) [inline runtime/onert/backend/train/MemoryPlanner.test.cc:34]: How about using `static_assert` to check invalid type in compile time?
```suggestion
  static_assert{false, "unimplemented to_index"};
```

File: runtime/onert/backend/train/MemoryPlanner.test.cc [hunk 2]
```diff
@@ -250,42 +341,232 @@ TEST(WICPlanner, claim_release_test)
...
+    p.verify(5, 1, 20, 0);
+
+    // VERIFY 6 - 20
+    p.verify(6, 1, 20, 20);
+
+    // CAPACITY - 40
+    p.capacity(40);
+  });
+}
+
```
Comment (zetwhite) [inline runtime/onert/backend/train/MemoryPlanner.test.cc:386]: After here, `layerscope_*` tests just copy-paste from the above test code.

File: runtime/onert/backend/train/MemoryPlanner.test.cc (hunk 1 above)
Comment (zetwhite) [reply runtime/onert/backend/train/MemoryPlanner.test.cc]: I tried, but `static_assert(false)` is not allowed in compile time.
It looks like if `false` is in `static_assert` compiler figured out invalid-program and it fails to compile :cry: 

```
/home/ONE_TWO/runtime/onert/backend/train/MemoryPlanner.test.cc: In function ‘T {anonymous}::to_index(uint32_t, uint32_t)’:
/home/ONE_TWO/runtime/onert/backend/train/MemoryPlanner.test.cc:34:17: error: static assertion failed: unimplemented to_index<T>()
   34 |   static_assert(false, "unimplemented to_index<T>()");
      |                 
```

File: runtime/onert/backend/train/MemoryPlanner.test.cc [hunk 3]
```diff
@@ -250,42 +342,47 @@ TEST(WICPlanner, claim_release_test)
...
+    // VERIFY 6 - 20
+    p.verify(6, 1, 20, 20);
+
+    // CAPACITY - 40
+    p.capacity(40);
+  });
 }
+
+// Add Testcase for LayerScopeTensorIndex, using PlannerVerifier
```
Comment (zetwhite) [inline runtime/onert/backend/train/MemoryPlanner.test.cc:388]: I'll add this at next PR.

File: runtime/onert/backend/train/MemoryPlanner.test.cc (hunk 1 above)
Comment (ragmani) [reply runtime/onert/backend/train/MemoryPlanner.test.cc:35]: Sorry for my wrong information :cry: 
Accroding to https://www.open-std.org/jtc1/sc22/wg21/docs/cwg_defects.html#941, you may be possible to apply deleted function templates to do it.
```cpp
template<typename T>
T to_index(uint32_t first, uint32_t second) = delete;
```

Comment (zetwhite) [reply runtime/onert/backend/train/MemoryPlanner.test.cc]: Wow, I didn't know that `=delete;` can be used for this case. Thank you for searching :+1: 

Updated

File: runtime/onert/backend/train/MemoryPlanner.test.cc [hunk 4]
```diff
@@ -25,9 +27,74 @@ using onert::ir::OperandIndex;
...
+  Planner<Index> _planner;
+};
+
+} // namespace
+
+TEST(BumpPlanner, disposable_claim_test)
+{
+  PlannerVerifier<BumpPlanner, DisposableTensorIndex> p;
+  /*
+  TODO - remove commented code
```
Comment (ragmani) [inline runtime/onert/backend/train/MemoryPlanner.test.cc:97]: (Optional)
Why didn't you remove the code?

File: runtime/onert/backend/train/MemoryPlanner.test.cc [hunk 5]
```diff
@@ -202,30 +286,35 @@ TEST(FirstFitPlanner, neg_release_twice)
...
+    // 0 RELEASE again
+    assert(on_only_debug_mode());
+  });
 }
 
-TEST(WICPlanner, claim_release_test)
+TEST(WICPlanner, disposable_claim_release_test)
 {
+  PlannerVerifier<WICPlanner, DisposableTensorIndex> p;
+  /*
```
Comment (ragmani) [inline runtime/onert/backend/train/MemoryPlanner.test.cc:317]: ditto

File: runtime/onert/backend/train/MemoryPlanner.test.cc [hunk 6]
```diff
@@ -15,8 +15,10 @@
  */
 
 #include <gtest/gtest.h>
+#include <stdexcept>
```
Comment (ragmani) [inline runtime/onert/backend/train/MemoryPlanner.test.cc:18]: This may be not necessary anymore.

**Review by ragmani — APPROVED** [review]
LGTM

File: runtime/onert/backend/train/MemoryPlanner.test.cc (hunk 4 above)
Comment (zetwhite) [reply runtime/onert/backend/train/MemoryPlanner.test.cc:97]: If I remove it, the diff page(https://github.com/Samsung/ONE/pull/13974/files) is messed up. 
I was afraid that checking - the test values have not changed at all - becomes difficult.

File: runtime/onert/backend/train/MemoryPlanner.test.cc (hunk 6 above)
Comment (zetwhite) [inline runtime/onert/backend/train/MemoryPlanner.test.cc:18]:
```suggestion
```

File: runtime/onert/backend/train/MemoryPlanner.test.cc [hunk 7]
```diff
@@ -17,6 +17,7 @@
 #include <gtest/gtest.h>
 
 #include "DisposableTensorIndex.h"
+#include "LayerScopeTensorIndex.h"
```
Comment (zetwhite) [inline runtime/onert/backend/train/MemoryPlanner.test.cc:20]:
```suggestion
```

**Review by ragmani — APPROVED** [review]

**Review by jyoungyun — APPROVED** [review]
LGTM

**Pull request merged**